src/Modules/UngradedModules.jl

```diff
@@ -0,0 +1,2689 @@
+    i += 1
+  end
+end
+
+# a basis of the free module 
+function basis(F::FreeMod)
+  bas = elem_type(F)[]
+  for i=1:dim(F)
+    s = Hecke.sparse_row(F.R, [(i, F.R(1))])
+    push!(bas, FreeModuleElem(s, F))
+  end
+  return bas
+end
+gens(F::FreeMod) = basis(F)
+
+# the ith generator of the free module
+function gen(F::FreeMod, i::Int)
+  @assert 0 < i <= ngens(F)
+  s = Hecke.sparse_row(F.R, [(i, F.R(1))])
+  return FreeModuleElem(s, F)
+end
+
+function Base.getindex(F::FreeMod, i::Int)
+  i == 0 && return zero(F)
+  return gen(F, i)
+end
+
+# the base ring of the free module
+base_ring(F::FreeMod) = F.R
+
+#TODO: Parent - checks everywhere!!!
+
+# the negative of a free module element
+-(a::FreeModuleElem) = FreeModuleElem(-a.coords, a.parent)
+
+# function to check whether a free module element is in a particular free module
+function check_parent(a::FreeModuleElem, b::FreeModuleElem)
+  if parent(a) !== parent(b)
+    error("elements not compatible")
+  end  
+end
+
+# Addition of free module elements
+function +(a::FreeModuleElem, b::FreeModuleElem)
+   check_parent(a, b)
+   return FreeModuleElem(a.coords+b.coords, a.parent)
+end
+
+# Subtraction of free module elements
+function -(a::FreeModuleElem, b::FreeModuleElem)
+    check_parent(a,b)
+    return FreeModuleElem(a.coords-b.coords, a.parent)
+end
+
+# Equality of free module elements
+function ==(a::FreeModuleElem, b::FreeModuleElem) 
+    check_parent(a,b)
+    return a.coords == b.coords
+end
+
+# scalar multiplication with polynomials, integers
+*(a::MPolyElem_dec, b::FreeModuleElem) = FreeModuleElem(a*b.coords, b.parent)
+*(a::MPolyElem, b::FreeModuleElem) = FreeModuleElem(a*b.coords, b.parent)
+*(a::Int, b::FreeModuleElem) = FreeModuleElem(a*b.coords, b.parent)
+*(a::Integer, b::FreeModuleElem) = FreeModuleElem(b.parent.R(a)*b.coords, b.parent)
+*(a::fmpq, b::FreeModuleElem) = FreeModuleElem(b.parent.R(a)*b.coords, b.parent)
+# zero of a free module
+zero(F::FreeMod) = FreeModuleElem(sparse_row(F.R, Tuple{Int, elem_type(F.R)}[]), F)
+# parent of a free module element
+parent(a::FreeModuleElem) = a.parent
+# test whether a free module element is zero
+iszero(a::FreeModuleElem) = Hecke.iszero(a.coords)
+
+# data structure for a generating systems for submodules
+# contains structures for the generators, the corresponding module on the Singular side, 
+# the embedding free module, the embedding free module on the Singular side
+# subquotients will be built from a tuple of submodules which again are given by 
+# generating sets. In this way, the Singular stuff is hidden on the higher structures
+# and all the conversion is taken care of here
+# a module generating system is generated from an array of free module elements
+# the fields are called O,S,F,SF rename?
+#
+# The same could be done rather on the level of vectors, that might be preferable if 
+# performance is ok.
+#
+mutable struct ModuleGens{T}
+  O::Array{FreeModuleElem{T}, 1}
+  S::Singular.smodule
+  F::FreeMod
```

Comment:
Should this be this?
```suggestion
  F::FreeMod{T}
```
If not, what is the purpose of `T` and how is it related to the the type parameter of `F`, if at all?